repose-aggregator/components/filters/herp/src/main/scala/org.openrepose.filters.herp/HerpFilter.scala

```diff
@@ -0,0 +1,134 @@
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.JavaConverters._
+
+class HerpFilter extends Filter with HttpDelegationManager with UpdateListener[HerpConfig] with LazyLogging {
+  private final val DEFAULT_CONFIG = "highly-efficient-record-processor.cfg.xml"
+
+  private var configurationService: ConfigurationService = _
+  private var config: String = _
+  private var initialized = false
+  private var herpLogger: Logger = _
+  private var serviceCode: String = _
+  private var region: String = _
+  private var dataCenter: String = _
+  private var mustacheTemplate: Mustache = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("HERP filter initializing ...")
+    config = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info("Initializing filter using config " + config)
+    val powerApiContext = ServletContextHelper.getInstance(filterConfig.getServletContext).getPowerApiContext
+    configurationService = powerApiContext.configurationService
+    val xsdURL: URL = getClass.getResource("/META-INF/schema/config/highly-efficient-record-processor.xsd")
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      config,
+      xsdURL,
+      this,
+      classOf[HerpConfig]
+    )
+
+    logger.trace("HERP filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    if (!initialized) {
+      logger.error("HERP filter has not yet initialized...")
+      servletResponse.asInstanceOf[HttpServletResponse].sendError(500)
+    } else {
+      logger.trace("HERP filter passing request...")
+      filterChain.doFilter(servletRequest, servletResponse)
+
+      logger.trace("HERP filter handling response...")
+      handleResponse(servletRequest.asInstanceOf[HttpServletRequest], servletResponse.asInstanceOf[HttpServletResponse])
+
+      logger.trace("HERP filter returning response...")
+    }
+  }
+
+  private def handleResponse(httpServletRequest: HttpServletRequest,
+                             httpServletResponse: HttpServletResponse) = {
+    def translateParameters(): Map[String, Array[Parameter]] = {
+      Option(httpServletRequest.getAttribute("http://openrepose.org/queryParams")) match {
+        case Some(parameters) => {
+          val parametersMap = parameters.asInstanceOf[java.util.Map[String, Array[String]]].asScala
+          parametersMap.mapValues((values :Array[String]) => values.map(Parameter(_))).toMap
+        }
+        case None => Map[String, Array[Parameter]]()
+      }
+    }
+
+    val templateValues = Map(
+      "userName" -> httpServletRequest.getHeader(OpenStackServiceHeader.USER_NAME.toString),
+      "impersonatorName" -> httpServletRequest.getHeader(OpenStackServiceHeader.IMPERSONATOR_NAME.toString),
+      "tenantID" -> httpServletRequest.getHeader(OpenStackServiceHeader.TENANT_ID.toString),
```

Comment:
Shouldn't the projectID be mapped here in case we're logging Keystone V3 requests?

Also, to parrot @adrianjgeorge, we should probably log all tenant/project IDs and not just the first to be returned.